File: day01/Solution.scala
package day01

import scala.io.Source

case object Solution extends App {

  def inverseCaptcha(input: String, skip: Int): Int = {
    val numbersToSum = for {
      i <- 0 until input.length
      if input.charAt(i) == input.charAt((i + skip) % input.length)
    } yield input.charAt(i).asDigit
    numbersToSum.sum
  }

  def partOne(input: String): Int = inverseCaptcha(input, 1)
  def partTwo(input: String): Int = inverseCaptcha(input, input.length /2)

  val input = Source.fromFile(getClass.getPackage.getName + "/input.in").mkString

  println(partOne(input))
  println(partTwo(input))

}

File: Day01/Solution.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace AdventOfCode2017.Day01 {

    class Solution : Solver {

        public string GetName() => "Inverse Captcha"; 
        
        public IEnumerable<object> Solve(string input) {
            yield return PartOne(input);
            yield return PartTwo(input);
        }

        int PartOne(string input) => InverseCaptcha(input, 1);

        int PartTwo(string input) => InverseCaptcha(input, input.Length / 2);

        int InverseCaptcha(string input, int skip) {
            return (
                from i in Enumerable.Range(0, input.Length)
                where input[i] == input[(i + skip) % input.Length]
                select int.Parse(input[i].ToString())
            ).Sum();
        }
    }
}
